examples/prediction/category_prediction/category.py

```diff
@@ -0,0 +1,188 @@
+# -*- coding: utf-8 -*-
+# ----------------------------------------------------------------------
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2013-2018, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+"""
+This example shows how to use the `CategoryEncoder` with `HTMPredictionModel`.
+
+This model uses the [MSNBC.com Anonymous Web Data Data Set][1] provided by
+[UCI Machine Learning Repository][2] to predict the next page the user is more
+likely to click. Each page is assigned a category and the user behavior is
+recorded as navigating from one page category to another. The dataset contains
+one user session per line. See [dataset][1] description for more information.
+
+For the purpose of this demonstration, the model will be trained on the
+first 100 user sessions and try to predict a random session from the learned
+sessions.
+
+References:
+  1: https://archive.ics.uci.edu/ml/datasets/MSNBC.com+Anonymous+Web+Data
+  2: Lichman, M. (2013). UCI Machine Learning Repository [http://archive.ics.uci.edu/ml].
+     Irvine, CA: University of California, School of Information and Computer Science
+"""
+import os
+import random
+import sys
+import zipfile
+from operator import itemgetter
+
+from nupic.frameworks.opf.model_factory import ModelFactory
+
+# List of categories in the dataset
+CATEGORIES = {
+  "1": "frontpage", "2": "news", "3": "tech", "4": "local", "5": "opinion",
+  "6": "on-air", "7": "misc", "8": "weather", "9": "msn-news", "10": "health",
+  "11": "living", "12": "business", "13": "msn-sports", "14": "sports",
+  "15": "summary", "16": "bbs", "17": "travel"
+}
+
+# HTM Prediction Model Parameters
+MODEL_PARAMS = {
+  "model": "HTMPrediction",
+  "version": 1,
+  "predictAheadTime": None,
+
+  "modelParams": {
+    "inferenceType": "TemporalMultiStep",
+    "sensorParams": {
+      "verbosity": 0,
+      "encoders": {
+
+        # Category encoder used for web page categories
+        "category": {
+          "fieldname": u"category",
+          "name": u"category",
+          "type": "CategoryEncoder",
+          "categoryList": CATEGORIES.keys(),
```

Comment:
Do we need to even specify the categories ahead of time? I thought the encoder can add in categories incrementally.